src/quicklist.c

```diff
@@ -448,35 +448,43 @@ REDIS_STATIC void _quicklistInsertNodeAfter(quicklist *quicklist,
     __quicklistInsertNode(quicklist, old_node, new_node, 1);
 }
 
-REDIS_STATIC int
-_quicklistNodeSizeMeetsOptimizationRequirement(const size_t sz,
-                                               const int fill) {
-    if (fill >= 0)
-        return 0;
+#define sizeMeetsSafetyLimit(sz) ((sz) <= SIZE_SAFETY_LIMIT)
 
-    size_t offset = (-fill) - 1;
-    if (offset < (sizeof(optimization_level) / sizeof(*optimization_level))) {
-        if (sz <= optimization_level[offset]) {
-            return 1;
-        } else {
-            return 0;
-        }
+/* Calculate the size limit or length limit of the quicklist node
+ * based on 'fill', and is also used to limit list listpack. */
+void quicklistNodeLimit(int fill, size_t *size, unsigned int *count) {
+    *size = SIZE_MAX;
+    *count = UINT_MAX;
+
+    if (fill >= 0) {
+        /* Ensure that one node have at least one entry */
+        *count = (fill == 0) ? 1 : fill;
     } else {
-        return 0;
+        size_t offset = (-fill) - 1;
+        size_t max_level = sizeof(optimization_level) / sizeof(*optimization_level);
+        if (offset >= max_level) offset = max_level - 1;
+        *size = optimization_level[offset];
     }
 }
 
-#define sizeMeetsSafetyLimit(sz) ((sz) <= SIZE_SAFETY_LIMIT)
+/* Check if the limit of the quicklist node has been reached to determine if
+ * insertions, merges or other operations that would increase the size of
+ * the node can be performed. */
+int quicklistNodeMeetLimit(int fill, size_t new_sz, unsigned int new_count) {
+    size_t sz_limit;
+    unsigned int count_limit;
+    quicklistNodeLimit(fill, &sz_limit, &count_limit);
 
-/* Calculate the size limit or length limit of the quicklist node
- * based on 'fill', and is also used to limit list listpack. */
-void quicklistSizeAndCountLimit(int fill, size_t *size, unsigned long *count) {
-    *size = SIZE_MAX;
-    *count = ULONG_MAX;
-    if (fill >= 0)
-        *count = fill;
+    if (likely(sz_limit != SIZE_MAX && new_sz <= sz_limit))
+        return 0;
+    /* when we return 0 above we know that the limit is a size limit (which is
+     * safe, see comments next to optimization_level and SIZE_SAFETY_LIMIT) */
+    else if (!sizeMeetsSafetyLimit(new_sz))
+        return 1;
+    else if (count_limit != UINT_MAX && new_count <= count_limit)
+        return 0;
     else
-        *size = optimization_level[-fill - 1];
+        return 1; 
 }
```

Comment:
* should that function now be named "exceeds" instead of "meets"? i.e. it seems you negated it and it returns 1 when we cant append to the node.
* i think we can drop the `else` since each `if` also returns.
* since we know (and even assume) that the size and count limits are mutually exclusive (exactly one of them isn't set to MAX), we can just check which of them is different than MAX, and then do `return new_sz <= sz_limit`
* the check about SafetyLimit is only needed if the limit is count based, and we're below the count (i.e. only if we have a chance to allow the insertion).
* bottom line, i think this function can be rewritten to be more readable (and renamed)

Reply:
Done